packages/store/src/Buffer.sol

```diff
@@ -0,0 +1,368 @@
+
+  /**
+   * @dev Appends _dataLength of the given bytes32 to the buffer (checking for overflows)
+   */
+  function append(
+    Buffer self,
+    bytes32 data,
+    uint128 _dataLength
+  ) internal pure {
+    uint128 _newLength = length(self) + _dataLength;
+    checkCapacity(self, _newLength);
+    appendUnchecked(self, data, _dataLength);
+  }
+
+  /**
+   * @dev Appends _dataLength of the given bytes32 to the buffer (without checking for overflows)
+   */
+  function appendUnchecked(
+    Buffer self,
+    bytes32 data,
+    uint128 _dataLength
+  ) internal pure {
+    uint256 _ptr = ptr(self);
+    uint128 _length = length(self);
+
+    // Update the current buffer length
+    _setLengthUnchecked(self, _length + _dataLength);
+
+    // Copy over given data to the buffer
+    assembly {
+      mstore(add(_ptr, _length), data)
+    }
+  }
+
+  /**
+   * @dev Appends the given bytes to the buffer (without checking for overflows)
+   */
+  function appendUnchecked(Buffer self, bytes memory data) internal pure {
+    uint256 _ptr = ptr(self);
+    uint128 _dataLength = uint128(data.length);
+    uint128 _length = length(self);
+    uint128 _newLength = _length + uint128(data.length);
+
+    // Update the current buffer length
+    _setLengthUnchecked(self, _newLength);
+
+    // Copy over given data to the buffer
+    assembly {
+      for {
+        let i := 0
+      } lt(i, _dataLength) {
+        i := add(i, 32)
+      } {
+        mstore(add(add(_ptr, _length), i), mload(add(data, add(32, i))))
+      }
+    }
+  }
+
+  /**
+   * @dev returns the buffer as a bytes array (without copying the data)
+   */
+  function toBytes(Buffer self) internal pure returns (bytes memory _bytes) {
+    uint256 _ptr = ptr(self);
+    assembly {
+      _bytes := sub(_ptr, 32)
+    }
+  }
+
+  function slice(
+    Buffer self,
+    uint256 _start,
+    uint256 _length
+  ) internal pure returns (bytes memory) {
+    return sliceBytes(toBytes(self), _start, _length);
+  }
+
+  /**
+   * @dev copies a slice of the buffer to a new memory location and returns it as a bytes array (without checking for overflows)
+   */
+
+  /**
+   * @dev copies the buffer contents to a new memory location and lays it out like a right aligned (eg uint) memory array with the given size per element (in bytes)
+   * @return arrayPtr the pointer to the start of the array, needs to be casted to the expected type using assembly
+   */
+  function toArray(Buffer self, uint256 elementSize) internal pure returns (uint256 arrayPtr) {
+    return toArray(self, elementSize, false);
+  }
+
+  /**
+   * @dev copies the buffer contents to a new memory location and lays it out like a memory array with the given size per element (in bytes)
+   * @return arrayPtr the pointer to the start of the array, needs to be casted to the expected type using assembly
+   */
+  function toArray(
+    Buffer self,
+    uint256 elementSize,
+    bool leftAligned
+  ) internal pure returns (uint256 arrayPtr) {
```

Comment:
wdyt about using `bytes32[] memory array` instead of `uint256 arrayPtr`? Cast will also need to be changed, but otherwise everything including gas (just tested) stays the same, and we avoid using pointers more than necessary.
It'd also be clearer that `toArray` can't make `bytes[]`/`string[]` arrays

Reply:
agree with this change!